internal/promote/storage.go

```diff
@@ -0,0 +1,410 @@
+package promote
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"sort"
+
+	"github.com/Masterminds/semver"
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/util"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-package/internal/github"
+)
+
+const (
+	remoteName = "elastic"
+
+	snapshotPackage = "snapshot"
+	stagingPackage  = "staging"
+	repositoryURL   = "https://github.com/%s/package-storage"
+)
+
+type fileContents map[string][]byte
+
+// PackageRevision represents a package revision stored in the package-storage.
+type PackageRevision struct {
+	Name    string
+	Version string
+
+	semver semver.Version
+}
+
+func (pr *PackageRevision) path() string {
+	return filepath.Join("packages", pr.Name, pr.Version)
+}
+
+// String method returns a string representation of the PackageRevision.
+func (pr *PackageRevision) String() string {
+	return fmt.Sprintf("%s-%s", pr.Name, pr.Version)
+}
+
+// PackageRevisions is an array of PackageRevision.
+type PackageRevisions []PackageRevision
+
+// Strings method returns an array of string representations.
+func (prs PackageRevisions) Strings() []string {
+	var entries []string
+	for _, pr := range prs {
+		entries = append(entries, pr.String())
+	}
+	return entries
+}
+
+// CloneRepository method clones the repository and changes branch to stage.
+func CloneRepository(stage string) (*git.Repository, error) {
+	r, err := git.Clone(memory.NewStorage(), memfs.New(), &git.CloneOptions{
+		URL:           fmt.Sprintf(repositoryURL, "elastic"),
+		RemoteName:    remoteName,
+		ReferenceName: plumbing.NewBranchReferenceName(stage),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "cloning package-storage repository failed")
+	}
+
+	err = r.Fetch(&git.FetchOptions{
+		RemoteName: remoteName,
+		RefSpecs:   []config.RefSpec{"refs/*:refs/*", "HEAD:refs/heads/HEAD"},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "fetch remote branches failed")
+	}
+
+	user, err := User(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading Git user failed")
+	}
+
+	_, err = r.CreateRemote(&config.RemoteConfig{
+		Name: user,
+		URLs: []string{
+			fmt.Sprintf(repositoryURL, user),
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "creating remote failed")
+	}
+	return r, nil
+}
+
+// ListPackages method lists available packages in the package-storage.
+// It skips technical packages (snapshot, staging).
```

Comment:
What do you mean by "technical" packages?

Reply:
It's a bit domain driven design. Packages like `snapshot` and `staging` don't below to the domain, but are strictly technical (we use them to highlight the currently used stage). 

If it's not straightforward I can reword it to something more clear.

Reply:
Yeah, "technical" is still unclear to me personally but that might be just me. Let's get another opinion on it. @ruflin WDYT?

Reply:
I can rephrase it:

`It skips packages: snapshot, staging`.

Reply:
Ok, I changed it to the above form.